design/ceph/admission-controller.md

```diff
@@ -0,0 +1,219 @@
+
+## Setup in Rook
+
+### Certificate Management
+
+#### Development and Testing 
+
+![Dev_AC_Image](./ac_dev_design.png)
+
+We will be opting for using for self-signed certificates approved by the kubernetes cluster itself. Once common.yaml and the operator has been deployed.  Using scripts we will be performing the following tasks
+* Creating the private key and certs using openssl.
+
+**Sample : Generating the public and private keys** 
+
+```
+openssl genrsa -out ${PRIVATE_KEY_NAME}.pem 2048
+ openssl req -new -key ${PRIVATE_KEY_NAME}.pem -subj "/CN=${service}.${namespace}.svc"
+
+#Created after CSR is approved
+openssl base64 -d -A -out ${PUBLIC_KEY_NAME}.pem 
+```
+
+* Creating and sending a CSR to kubernetes for approval
+
+**Sample : Certificate Signing Request (CSR) in Kubernetes**
+```
+apiVersion: certificates.k8s.io/v1beta1
+kind: CertificateSigningRequest
+metadata:
+  name: ${csrName}
+spec:
+  request: $(cat server.csr | base64 | tr -d '\n')
+  usages:
+  - digital signature
+  - key encipherment
+  - server auth
+```
+Verify with
+` kubectl get csr ${csrName}`
+
+**Sample : Approval of Signed Certificate**
+
+```
+kubectl certificate approve ${csrName}
+```
+
+If it is approved, then we can check the certificate with following command 
+
+```
+$(kubectl get csr ${csrName} -o jsonpath='{.status.certificate}')
+```
+
+* Once approved, a generic secret will be created with the given public and private key which will be later mounted onto the server pod for use.
+
+**Sample : Creating a Secret in Kubernetes**
+```
+kubectl create secret generic ${secret} \
+        --from-file=key.pem=${PRIVATE_KEY_NAME}.pem \
+        --from-file=cert.pem=${PUBLIC_KEY_NAME}.pem
+```
+* Modifying the webhook config to inject CA bundle onto the ValidatingWebhookConfig
+* All the above resources will be created in rook-ceph namespace
+
+
+Using the above approach, the dev/admins will have the responsibility of rotating the certificates when they expire.
+
+Below is an excerpt of what a ValidatingWebhookConfig looks like 
+
+```
+apiVersion: admissionregistration.k8s.io/v1beta1
+kind: ValidatingWebhookConfiguration
+metadata:
+  name: demo-webhook
+webhooks:
+  - name: webhook-server.webhook-demo.svc
+    clientConfig:
+      service:
+        name: webhook-server
+        namespace: rook-ceph
+        path: "/validate"
+      caBundle: ${CA_PEM_B64}
+    rules:
+      - operations: [ "CREATE" ]
+        apiGroups: [""]
+        apiVersions: ["v1"]
+        resources: ["pods"]
+```
+
+We can make changes to the above values according to intercept based on whether a resource is being updated/deleted/created or change the type of resource or the request path which will be sent to the server.
+
+#### Production
+
+The onus of providing proper certificates and CA bundle for the server would be on the user. The following example would provide how it would work in an OCP cluster environment
+
+##### Openshift Container Platform
```

Comment:
This section looks like a candidate for Documentation and not design doc.

Reply:
@umangachapagain What changes do you suggest for this ?

Reply:
I have segregated the PR into 2 commits. One commit is for the design documentation and the 2nd commit adds docuemntation for deploying in OCP